File: sdk/js/src/sdk/signatures/sign-withdraw-from-lock.ts
import type { WalletClient } from 'viem'
import type { Address } from '../types'
import {
  createDomain,
  WITHDRAW_FROM_LOCK_TYPES,
  type WithdrawFromLockMessage,
} from './eip712-types'

export interface SignWithdrawFromLockParams {
  walletClient: WalletClient
  chainId: number
  verifyingContract: Address
  message: WithdrawFromLockMessage
}

export async function signWithdrawFromLockMessage({
  walletClient,
  chainId,
  verifyingContract,
  message,
}: SignWithdrawFromLockParams): Promise<`0x${string}`> {
  const account = walletClient.account
  if (!account) {
    throw new Error('No account connected to wallet client')
  }

  const domain = createDomain(chainId, verifyingContract)

  const signature = await walletClient.signTypedData({
    account,
    domain,
    types: WITHDRAW_FROM_LOCK_TYPES,
    primaryType: 'WithdrawFromLock',
    message,
  })

  return signature
}

File: sdk/py/src/flexvaults/signatures/sign_withdraw_from_lock.py
from __future__ import annotations

from dataclasses import dataclass

from eth_account.signers.local import LocalAccount

from ..types import Address, HexString, Network
from ._signer import sign_typed_data
from .eip712_types import (
    WITHDRAW_FROM_LOCK_TYPES,
    WithdrawFromLockMessage,
    create_domain,
)


@dataclass
class SignWithdrawFromLockParams:
    account: LocalAccount
    network: Network
    verifying_contract: Address
    message: WithdrawFromLockMessage


def sign_withdraw_from_lock_message(params: SignWithdrawFromLockParams) -> HexString:
    domain = create_domain(params.network, params.verifying_contract)

    message_data = {
        "userAddress": params.message.user_address,
        "toAddress": params.message.to_address,
        "lockId": params.message.lock_id,
        "amount": params.message.amount,
        "nonce": params.message.nonce,
    }

    return sign_typed_data(
        account=params.account,
        domain=domain,
        types=WITHDRAW_FROM_LOCK_TYPES,
        primary_type="WithdrawFromLock",
        message=message_data,
    )
